hathor/p2p/sync_mempool.py

```diff
@@ -0,0 +1,173 @@
+        tx = yield self._get_data(tx_id)
+        # TODO Verify tx.
+        # tx.verify()
+        return tx
+
+    def _get_data(self, tx_id: bytes) -> Deferred:
+        """Internal method to request the tx."""
+        assert self._download_tx_deferred is None
+        assert self._download_tx_hash is None
+        self._download_tx_deferred = Deferred()
+        self._download_tx_hash = tx_id
+        self.sync_manager.send_get_data(tx_id, origin='mempool')
+        return self._download_tx_deferred
+
+    def on_new_mempool_tx(self, tx: BaseTransaction) -> None:
+        """Called when a requested tx is received."""
+        assert self._download_tx_deferred is not None
+        if tx.hash != self._download_tx_hash:
+            # Peer sent the wrong transaction?!
+            # Maybe we should ban the peer.
+            raise Exception('This should never happen.')
+        deferred = self._download_tx_deferred
+        self._download_tx_deferred = None
+        self._download_tx_hash = None
+        deferred.callback(tx)
+
+    def run(self) -> None:
+        """Run a loop of the sync-v2 mempool and schedules the next call.
+        It can safely be called multiple times.
+        """
+        self.run_once()
+        self.reactor.callLater(1000, self.run)
+
+    @inlineCallbacks
+    def run_once(self) -> Generator[None, BaseTransaction, None]:
+        """Run a single loop of the sync-v2 mempool."""
+        if not self.missing_tips:
+            # No missing tips? Let's get them!
+            self.sync_manager.send_get_tips()
+
+        elif not self.deque:
+            # Not downloading? Let's start!
+            tx_id = next(iter(self.missing_tips))
+            tx = yield self.get_tx(tx_id)
+            self._dfs_start(tx)
+
+        else:
+            # Nothing to do. \o/
+            pass
+
+    def on_new_tips(self, tx_hashes: List[bytes]) -> None:
+        """Called when a new list of tips is received from the peer."""
+        for tx_id in tx_hashes:
+            if self.tx_storage.transaction_exists(tx_id):
+                continue
+            self.missing_tips.add(tx_id)
+
+    def _dfs_start(self, tx: BaseTransaction) -> None:
+        """Initialize and start the DFS."""
+        assert len(self.deque) == 0
+        self.deque.append(tx)
+        self._dfs()
+
+    def get_one_missing_dep(self, tx: BaseTransaction) -> Optional[bytes]:
+        """Get this first missing dependency found of tx."""
+        assert not tx.is_block
+        for txin in tx.inputs:
+            if not self.tx_storage.transaction_exists(txin.tx_id):
+                return txin.tx_id
+        for parent in tx.parents:
+            if not self.tx_storage.transaction_exists(parent):
+                return parent
+        return None
+
+    def add_tx(self, tx: BaseTransaction) -> None:
+        """Add tx to the DAG."""
+        assert tx.hash is not None
+        self.missing_tips.discard(tx.hash)
+        self.manager.on_new_tx(tx)
```

Comment:
Why if it fails?

Reply:
_What_ if it fails? It will raise an exception and `_dfs` will crash (the function will exit and its deferred will be rejected, which isn't handled anywhere). Maybe this should be handled on `_dfs`.

Reply:
_What_ if it fails? It will raise an exception and `_dfs` will crash (the function will exit and its deferred will be rejected, which isn't handled anywhere). Maybe this should be handled on `_dfs`.

Reply:
And it will run again when `NodeBlockSync.run_sync` is called again from the `LoopingCall`.